File: src/main/java/com/vassant/kata/domain/Balance.java
package com.vassant.kata.domain;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
@AllArgsConstructor(access = AccessLevel.PRIVATE)
public final class Balance {

    private int value;

    static Balance of(int value) {
        return new Balance(value);
    }

    Balance deposit(Amount amount) {
        return Balance.of(value + amount.getValue());
    }

    Balance withdraw(Amount amount) {
        return Balance.of(value - amount.getValue());
    }

    boolean hasEnoughSavings(Amount amount) {
        return value >= amount.getValue();
    }
}

File: src/main/kotlin/com/vassant/kata/domain/Balance.kt
package com.vassant.kata.domain

data class Balance(private val value: Int) {
    infix fun deposit(amount: Amount): Balance = Balance(value + amount)

    infix fun withdraw(amount: Amount): Balance = Balance(value - amount)

    fun hasEnoughSavings(amount: Amount): Boolean = value >= amount.value

    infix fun apply(operation: Operation): Balance =
            when (operation) {
                is Withdraw -> this withdraw operation.amount
                is Deposit -> this deposit operation.amount
            }
}

operator fun Int.plus(amount: Amount): Int = this + amount.value
operator fun Int.minus(amount: Amount): Int = this - amount.value